File: jvm/src/main/java/example/Compiler.java
package example;

import kand.asm.*;

import java.io.FileOutputStream;
import java.io.IOException;

public class Compiler {
    public static void writeFile(String filename, byte[] data) throws IOException {
        FileOutputStream out=new FileOutputStream(filename);
        out.write(data);
        out.close();
    }

    public static void main(String[] args) throws IOException {
        Method exp = new Method("kand/runtime/Core", "add",
                new Exp[] { new Nm(100), new Method("kand/runtime/Core", "add",
                        new Exp[] { new Var(1), new Var(2)})});

        writeFile("example/add.class", FnClassWriter.write("example/add", new String[] {"x", "y"}, exp));
        writeFile("example/main.class", MainClassWriter.write("example/main"));
    }
}

File: src/eta/Compiler.hs
{-# LANGUAGE TypeFamilies, DataKinds, ScopedTypeVariables #-}

module Compiler ( compile ) where

import qualified Data.ByteString as BS
import Asm.FnClassWriter as FnClass
import Asm.MainClassWriter as MainClass
import Asm.AsmExp
import Asm.AsmMethod
import Asm.AsmVar
import Asm.AsmNm
import Parser
import Java

toJStringArray :: [String] -> JStringArray
toJStringArray strs = toJava jstrings
  where jstrings = map toJava strs :: [JString]

compile :: [String] -> IO ()
compile (filename:[]) = do
  code <- byteArray
  BS.writeFile "example/add.class" (BS.pack (fromJava code))
  BS.writeFile "example/main.class" (BS.pack (fromJava $ MainClass.write "example/main"))
  return ()
  where
    byteArray = java $ do
      var1 <- newAsmVar 1
      var2 <- newAsmVar 2
      args <- arrayFromList [ (superCast var1 :: AsmExp)
                            , (superCast var2 :: AsmExp) ]
      exp <- newAsmMethod "kand/runtime/Core" "add" args
      return $ FnClass.write "example/add" (toJStringArray ["x", "y"]) exp
